posts/2018-07-12-servant-dsl-typelevel.md

```diff
@@ -0,0 +1,664 @@
+it gets as input. That is, we want a type that depends on a value, i.e
+dependent types.
+
+The alternative would be:
+
+``` haskell
+-- linkTo assumes that the capture values are given in the same order as we want
+-- them to appear in the url.
+linkTo :: Endpoint -> [String] -> Link
+linkTo (Static str rest) captureValues  = str : linkTo rest captureValues
+-- when there is at least one value left in the list, we use it.
+-- when there isn't... we error out.
+linkTo (Capture rest)    (c : cs)       = c   : linkTo rest cs
+linkTo (Capture rest)    []             = error "linkTo: capture value needed but the list is empty" -- :-(
+linkTo (Verb method)     _              = []
+```
+
+This solution is very unsatisfactory, first and foremost because it is possible
+for it to error out if we don't supply the right number of captures. However,
+it will also silently let us pass too many capture values without telling us
+that some of them are not used. Such a function should be total, we really don't
+want an implementation that can let us down if we're not very careful.
+
+Fortunately, GADTs can help here. We could turn `Endpoint` into
+a GADT that tracks captures and then use some type-level computations
+to get the type of the link-making function from our list of captures, as well
+as define the link making function through typeclass instances that would
+go through the captures and add an argument for each of them. Request bodies,
+query parameters, headers? We could probably track them too, in a similar way.
+Or we could unify it all by basically building up and tracking actual servant
+API types through a `GADT` version of Endpoint's type argument, and do some of
+what servant does at the type-level, with everything else at the value-level.
+
+However, all those approaches have a big problem. Once you've made a decision,
+it is set in stone, in a way. To be more precise, with all these approaches,
+if you want to extend your web app description language you need to add a new
+constructor to your `Endpoint` type, and you then need to handle this new
+constructor in all you functions that pattern match on `Endpoint` constructors
+since they've instantly become partial. Not to mention that just the act of
+adding a constructor requires rebuilding the entire library. You cannot explore
+two different directions simultaneously without breaking code, you cannot add
+new constructs you hadn't thought of without touching the library, e.g just
+locally in a project of yours. Extensibility and modularity
+were central requirements as we had been bitten by the lack of them
+in libraries that we were using at the time.
+
+So... how do people build extensible/modular DSLs in Haskell? The next section
+talks about the general problem behind this and a solution that I read about
+that gets us halfway to servant.
+
+# The Expression Problem
+
+To quote [Phil Wadler](http://homepages.inf.ed.ac.uk/wadler/papers/expression/expression.txt):
```

Comment:
Maybe Philip is better?

Reply:
I always see "Phil Wadler" everywhere, and I'm not much of a formal person, so I think I'd leave it like this.